source/extensions/clusters/dynamic_forward_proxy/cluster.h

```diff
@@ -90,19 +87,14 @@ class Cluster : public Upstream::BaseDynamicClusterImpl,
     Cluster& cluster_;
   };
 
-  HostInfoMapSharedPtr getCurrentHostMap() {
-    absl::ReaderMutexLock lock(&host_map_lock_);
-    return host_map_;
-  }
-
   void
-  addOrUpdateWorker(const std::string& host,
-                    const Extensions::Common::DynamicForwardProxy::DnsHostInfoSharedPtr& host_info,
-                    std::shared_ptr<HostInfoMap>& new_host_map,
-                    std::unique_ptr<Upstream::HostVector>& hosts_added);
-  void swapAndUpdateMap(const HostInfoMapSharedPtr& new_hosts_map,
-                        const Upstream::HostVector& hosts_added,
-                        const Upstream::HostVector& hosts_removed);
+  addOrUpdateHost(const std::string& host,
+                  const Extensions::Common::DynamicForwardProxy::DnsHostInfoSharedPtr& host_info,
+                  std::unique_ptr<Upstream::HostVector>& hosts_added)
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(host_map_lock_);
+  void updatePriorityState(const Upstream::HostVector& hosts_added,
```

Comment:
I think it's important to not hold the lock while you call `updateClusterPrioritySet` inside this function, since that can invoke a large number of callbacks, etc. Can you potentially refactor to drop the lock before calling out to all the callbacks? Or is holding the lock required somehow?

Reply:
I refactored the code in 7862a45 to release the lock before calling `updateClusterPrioritySet`.